### PR TITLE
Obj index update

### PR DESCRIPTION
**Pull request by movestill**
Lambda functions that support id indexing.

Accompanying PRs:
* https://github.com/jhuapl-boss/boss-manage/pull/8
* https://github.com/jhuapl-boss/boss/pull/10
* https://github.com/jhuapl-boss/spdb/pull/3

### CODE REVIEW

File: lambda/fanout_write_id_index_lambda.py [hunk 1]
```diff
@@ -0,0 +1,158 @@
...
+    }
+
+    num_execs = get_lambda_execs_in_last_min()
+    if max_execs - num_execs < len(fanout_subargs) * 2:
+        # Don't fanout during this execution.  Wait for more lambda 
+        # availability.
+        fanout_args['wait_secs'] = 10 + randint(15, 25)
+        print('Too many lambdas, pausing fanout for {} secs.'.format(
+            fanout_args['wait_secs']))
+        return fanout_args
```
Comment (dkleissa) [inline lambda/fanout_write_id_index_lambda.py]: managing lambda invocations manually seems fine, but has letting AWS manage this through reserved capacity been thought through?

File: lambda/index_find_cuboids_lambda.py [hunk 2]
```diff
@@ -0,0 +1,120 @@
...
+        'ExpressionAttributeValues': {':lookupKeyVal': {'S': key}},
+        'Limit': limit,
+        'ProjectionExpression': '#objKey, #versionNode'
+    }
+
+    if 'exclusive_start_key' in event and event['exclusive_start_key'] != '':
+        query_args['ExclusiveStartKey' ] = json.loads(event['exclusive_start_key'])
+
+
+    resp = dynamo.query(**query_args)
```
Comment (dkleissa) [inline lambda/index_find_cuboids_lambda.py]: check for 200/handle errors better here?

File: lambda/split_cuboids_lambda.py [hunk 3]
```diff
@@ -0,0 +1,129 @@
...
+    num_ids = int(event['num_ids'])
+    if num_ids == 0:
+        if 'sub_args' not in event or len(event['sub_args']) == 0:
+            # Nothing left to do.
+            output = {'finished': True}
+            return output
+
+    fanout_args = get_fanout_args(event)
+
+    # Change this key?
```
Comment (dkleissa) [inline lambda/split_cuboids_lambda.py]: comment is confusing. Is this a todo?

Comment (movestill) [reply lambda/split_cuboids_lambda.py]: Yeah, a potential todo if I start reserving lambda executions for each lambda involved in indexing.

File: lambda/fanout_write_id_index_lambda.py (hunk 1 above)
Comment (movestill) [reply lambda/fanout_write_id_index_lambda.py]: Yeah, but I think I still need to potentially slow things down here, so there's a greater chance of each step function succeeding w/o specifying a crazy number of retries. That being said, I have new code to push that does a "decorrelated" backoff. Step function retries don't allow you to specify a random element, so the current step functions all collide when trying to write the same id. Then they all wait about the same amount of time, and collide again! :)

File: lambda/index_find_cuboids_lambda.py (hunk 2 above)
Comment (movestill) [reply lambda/index_find_cuboids_lambda.py]: I'm just letting any exceptions be thrown and letting the step function handle the retry in this case.

**Pull request merged**